awx/main/tests/functional/task_management/test_scheduler.py

```diff
@@ -383,3 +383,37 @@ def test_job_not_blocking_inventory_update(default_instance_group, job_template_
         dependency_graph = DependencyGraph(None)
         dependency_graph.add_job(job)
         assert not dependency_graph.is_job_blocked(inventory_update)
+
+
+@pytest.mark.django_db
+def test_generate_dependencies_only_once(job_template_factory):
+    objects = job_template_factory('jt', organization='org1', project='proj')
+
+    job = objects.job_template.create_job()
+    project = objects.project
+    project.scm_update_on_launch = True
+    job.status = "pending"
+    job.name = "job_gen_dep"
+    job.save()
+    project.save()
+
+    with mock.patch("awx.main.scheduler.TaskManager.start_task"):
+        # job starts with dependencies_processed as False
+        assert not job.dependencies_processed
+        # run one cycle of ._schedule() to generate dependencies
+        TaskManager()._schedule()
+
+        # make sure dependencies_processed is now True
+        job = Job.objects.filter(name="job_gen_dep")[0]
+        assert job.dependencies_processed
+
+        # Run ._schedule() again, but make sure .generate_dependencies() is not
+        # called with job in the argument list
+        tm = TaskManager()
+        tm.generate_dependencies = mock.MagicMock()
+        tm._schedule()
+
+        # .call_args is tuple, (positional_args, kwargs), [0][0] then is
+        # the first positional arg, i.e. the first argument of
+        # .generate_dependencies()
+        assert job not in tm.generate_dependencies.call_args[0][0]
```

Comment:
Since there are no other jobs in the database here, I would prefer a stricter assertion of `tm.generate_dependencies.call_args[0][0] == []`

Reply:
@AlanCoding because project.scm_update_on_launch == True, then .call_args[0][0] contains [ProjectUpdate]. I like the idea of having a stricter assertion though, so I will turn off the project update and assert the empty list.